DREAM/Optimization/tsgGradientDescent.cpp

```diff
@@ -0,0 +1,128 @@
+    OptimizationStatus status;
+    status.performed_iterations = 0;
+    status.residual = std::numeric_limits<double>::max();
+
+    int num_dimensions = state.num_dimensions;
+    std::vector<double> &x = state.getXRef();
+    std::vector<double> x0(x), gx0(num_dimensions), gx(num_dimensions), z0(num_dimensions), xStep(num_dimensions);
+    double fx(func(x)), fx0, fxStep;
+    grad(x, gx);
+
+    while(status.performed_iterations < num_iterations) {
+        // Iteration swaps must be performed first (instead of last) to avoid reverting x to x0 after work has been completed.
+        std::swap(x0, x);
+        std::swap(fx0, fx);
+        std::swap(gx0, gx);
+        // Find the next stepsize/candidate point by making sure it satisfies the well-known descent inequality (see γ_u in the
+        // reference paper associated with this function).
+        double lhs = 0;
+        double rhs = 0;
+        do {
+            if (status.performed_iterations >= num_iterations) break;
+            for (int j=0; j<num_dimensions; j++)
+                z0[j] = x0[j] - gx0[j] * state.adaptive_stepsize;
+            proj(z0, xStep);
+            fxStep = func(xStep);
+            lhs += fxStep - fx0;
+            for (int j=0; j<num_dimensions; j++) {
+                double delta = xStep[j] - x0[j];
+                rhs += delta * delta / (2.0 * state.adaptive_stepsize);
+                lhs -= gx0[j] * delta;
+            }
+            state.adaptive_stepsize /= decrease_coeff;
+            status.performed_iterations++;
+        } while (lhs > rhs + TasGrid::Maths::num_tol);
+        std::swap(xStep, x);
+        std::swap(fxStep, fx);
+        state.adaptive_stepsize *= decrease_coeff; // offset the do-while loop.
+        // Check for approximate stationarity.
+        grad(x, gx);
+        status.residual = compute_stationarity_residual(x, x0, gx, gx0, state.adaptive_stepsize);
+        if (status.residual <= tolerance) break;
+        // Optimistic stepsize update (see γ_d in the referenced paper above).
+        state.adaptive_stepsize *= increase_coeff;
+    }
+    return status;
+    
+}
+
+OptimizationStatus GradientDescent(const ObjectiveFunctionSingle &func, const GradientFunctionSingle &grad,
+                                   const double increase_coeff, const double decrease_coeff, const int num_iterations,
+                                   const double tolerance, GradientDescentState &state) {
+    // Wrapper to the proximal version with projection function == identity function.
+    return GradientDescent(func, grad, identity, increase_coeff, decrease_coeff, num_iterations, tolerance, state);
+};
+
+OptimizationStatus GradientDescent(const GradientFunctionSingle &grad, const double stepsize, const int num_iterations,
+                                   const double tolerance, std::vector<double> &state) {
+    OptimizationStatus status;
+    status.performed_iterations = 0;
+    status.residual = std::numeric_limits<double>::max();
```

Comment:
I personally feel very uneasy when I operate so close to numeric instability, and in the current implementation (and the one that I propose below), this isn't necessary.
* currently, you're not doing any residual testing until you overwrite this, you're not using this value so it can be anything
* in the proposed (and current) solution, you can set:
```
status.residual = tolerance + 1.0;
```

Reply:
Changed using the suggestion below this one.